### PR TITLE
Fix gtest data race detected by tsan check

### PR DESCRIPTION
**Pull request by yibin87**
### What problem does this PR solve?

Issue Number: ref #8285 #8553

Problem Summary:
In #8553, mutex is added to protect prepareMPPTasks. However, the DAGRequestBuilder parameter is constructed outside the protection, and the context's index will be updated during the construction. This is caused by the two-phase compilation: first for operators without exchange, second for exchange operators. Thus adjust the mutex protection scope in this pr.

Set server's mock storage at beginning to avoid harmless data races.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Flash/tests/gtest_compute_server.cpp [hunk 1]
```diff
@@ -147,11 +147,16 @@ class ComputeServerRunner : public DB::tests::MPPTaskTestUtils
             BlockInputStreamPtr stream;
             try
             {
-                auto tasks = prepareMPPTasks(
```
Comment (SeaRise) [inline dbms/src/Flash/tests/gtest_compute_server.cpp:150]: how about `prepareMPPTasks([&] () { return context.xxx }, properties)`?
and then `DAGRequestBuilder` will under lock

Comment (SeaRise) [reply dbms/src/Flash/tests/gtest_compute_server.cpp:150]: Maybe `MockDAGRequestContext::index` should be atomic_size_t?

Comment (yibin87) [reply dbms/src/Flash/tests/gtest_compute_server.cpp:150]: Sounds great, I'll change it.

Comment (SeaRise) [reply dbms/src/Flash/tests/gtest_compute_server.cpp:150]: Maybe `thread_local size_t index`? but seems that mac support `thread_local` :(

**Review by SeaRise — APPROVED** [review]

**Review by windtalker — APPROVED** [review]
LGTM

**Pull request merged**